lib/hanami/action/flash.rb

```diff
@@ -6,35 +6,59 @@ module Action
     # @since 0.3.0
     # @api private
     class Flash
+
+      class Kept
+        attr_reader :data
+
+        def initialize(data)
+          @data = data
+          @request_count = 0
+        end
+
+        def find(key)
+          data[key]
+        end
+
+        def key?(key)
+          data.key?(key)
+        end
+
+        def increase_request_count!
+          @request_count = @request_count += 1
+        end
+
+        def expired?
+          @request_count >= 2
+        end
+      end
+
+      private_constant :Kept
       # Session key where the data is stored
       #
       # @since 0.3.0
       # @api private
       SESSION_KEY = :__flash
 
-      # Session key where the last request_id is stored
+      # Session key where keep data is store for redirect
       #
-      # @since 0.4.0
+      # @since x.x.x
       # @api private
-      LAST_REQUEST_KEY = :__last_request_id
+      KEPT_KEY = :__kept_key
 
       # Initialize a new Flash instance
       #
       # @param session [Rack::Session::Abstract::SessionHash] the session
-      # @param request_id [String] the HTTP Request ID
       #
       # @return [Hanami::Action::Flash] the flash
       #
       # @see http://www.rubydoc.info/gems/rack/Rack/Session/Abstract/SessionHash
       # @see Hanami::Action::Rack#session_id
-      def initialize(session, request_id)
+      def initialize(session)
         @session         = session
-        @request_id      = request_id
-        @last_request_id = session[LAST_REQUEST_KEY]
-        @merged          = {}
+        @keep            = false
 
-        session[SESSION_KEY]             ||= {}
-        session[SESSION_KEY][request_id] ||= {}
+        session[KEPT_KEY]                ||= []
```

Comment:
Some of these spaces and be removed. Might even be able to line these lines up with `@session` and `@keep`, if ya want